src/java.base/share/classes/java/lang/StableValue.java

```diff
@@ -0,0 +1,746 @@
+import java.util.Map;
+import java.util.NoSuchElementException;
+import java.util.Objects;
+import java.util.RandomAccess;
+import java.util.Set;
+import java.util.function.Function;
+import java.util.function.IntFunction;
+import java.util.function.Supplier;
+
+/**
+ * A stable value is a holder of shallowly immutable content that can be lazily computed.
+ * <p>
+ * A {@code StableValue<T>} is typically created using the factory method
+ * {@linkplain StableValue#of() {@code StableValue.of()}}. When created this way,
+ * the stable value is <em>unset</em>, which means it holds no <em>content</em>.
+ * Its content, of type {@code T}, can be <em>set</em> by calling
+ * {@linkplain #trySet(Object) trySet()}, {@linkplain #setOrThrow(Object) setOrThrow()},
+ * or {@linkplain #orElseSet(Supplier) orElseSet()}. Once set, the content
+ * can never change and can be retrieved by calling {@linkplain #orElseThrow() orElseThrow()}
+ * , {@linkplain #orElse(Object) orElse()}, or {@linkplain #orElseSet(Supplier) orElseSet()}.
+ * <p>
+ * A stable value that is <em>set</em> is treated as a constant by the JVM, enabling the
+ * same performance optimizations that are available for {@code final} fields.
+ * As such, stable values can be used to replace {@code final} fields in cases where
+ * <em>at-most-once</em> update semantics is crucial, but where the eager initialization
+ * semantics associated with {@code final} fields is too restrictive.
+ * <p>
+ * Consider the following example where a stable value field "{@code logger}" is a
+ * shallowly immutable holder of content of type {@code Logger} and that is initially
+ * created as <em>unset</em>, which means it holds no content. Later in the example, the
+ * state of the "{@code logger}" field is checked and if it is still <em>unset</em>,
+ * the content is <em>set</em>:
+ *
+ * {@snippet lang = java:
+ * public class Component {
+ *
+ *    // Creates a new unset stable value with no content
+ *    // @link substring="of" target="#of" :
+ *    private static final StableValue<Logger> logger = StableValue.of();
+ *
+ *    private Logger getLogger() {
+ *        if (!logger.isSet()) {
+ *            logger.trySet(Logger.create(Component.class));
+ *        }
+ *        return logger.orElseThrow();
+ *    }
+ *
+ *    public void process() {
+ *        getLogger().info("Process started");
+ *        // ...
+ *    }
+ * }
+ *}
+ * <p>
+ * Note that the content can only be set at most once.
+ * In the example above, the {@code logger} field is declared {@code static final} which
+ * is a prerequisite for being treated as a constant by the JVM.
+ *
+ * <p>
+ * To guarantee that, even under races, only one instance of {@code Logger} is ever
+ * created, the {@linkplain #orElseSet(Supplier) orElseSet()} method can be used
+ * instead, where the content is atomically and lazily computed via a
+ * {@linkplain Supplier supplier}. In the example below, the supplier is provided in the
+ * form of a lambda expression:
```

Comment:
I suspect that developers might use the aforementioned snippet without knowing that under a race multiple logger instances may be created (even if only one is then used). So I'd put a comment  above trySet indicating that.